misc: Add a generic cache template

pranith opened the pull request:
This allows us to use a shorter tag than otherwise possible.

Change-Id: Ib339efd97bf33b706f93eb7f0eb5851ceb51afb6

pranith committed:
misc: Add a generic cache library

Add a generic cache library modeled after AssociativeSet that can be used for
constructing internal caching structures.

Change-Id: I1767309ed01f52672b32810636a09142ff23242f

pranith committed:
cpu-o3: Use the generic cache library to build store sets

This allows us to build an associative store set structure
and also experiment with various replacement policies for
the store set.

Change-Id: I18ddad2346b82423c01ef9117a5d3d85f7039be7

pranith committed:
cpu: pred: Use cache library to build the BTB

This allows us to make the BTB associative and also experiment with various
replacement and indexing policies for the BTB.

Change-Id: Ie9085fc4728e000a9b62ed48a0ecd1a57a22201c

KaiBatley committed:
gpu-compute: WAX dependency detection

WAX Dependencies would be missed if a RAW Dependency also existed.

Change-Id: I2a9e50b9d0540a30de9c1bf6bb544c7b9654cb29

powerjg commented on src/mem/cache/tags/indexing_policies/base.hh:
```diff
@@ -96,6 +96,16 @@ class BaseIndexingPolicy : public SimObject
+    const int tagBits;
+
+    /**
+     * The tag mask.
```
Can you add more information here about how this is used?

powerjg commented on src/mem/cache/tags/indexing_policies/IndexingPolicies.py:
```diff
@@ -44,6 +44,9 @@ class BaseIndexingPolicy(SimObject):
     assoc = Param.Int(Parent.assoc, "associativity")
 
+    # number of bits in the tag
+    tag_bits = Param.Unsigned(64, "number of bits in the tag")
```
Can you improve the help string here and/or improve the name of the variable? There are a few confusions possible here.
- The number of bits in the tag would really be 64 - index bits - offset bits, so 64 bits for the tag is misleading.
- What would this parameter be used for? Why would you set the number of tag bits to not be *all* of the remaining bits? Adding this explanation to the help string will enable others to understand and use this feature.

pranith replied:
I used 64 so that I don't change the current semantics of the code. I am open to suggestions on what to use here, so please let me know.

There are cases where the tag is shortened either by hashing or dropping the upper bits that do not change often to save storage. BTB is one example which uses the reduced tags. I added a patch showing how the reduced tag is used in BTB in the series.

powerjg reviewed (commented):
The code seems OK, but before we merge this I would like to understand the use case. When would you set the tag to *not* be the entire address for an indexing policy?

If you could update the commit message with (1) the problem that you are solving and (2) a detailed explanation of why this is the solution to the problem, it will make it easier to review the code.

odanrc commented on src/mem/cache/tags/indexing_policies/base.hh:
```diff
@@ -96,6 +96,16 @@ class BaseIndexingPolicy : public SimObject
 
+    /**
+     * The number of tag bits per entry
+     */
```
Nit: This variable is not really needed

pranith replied:
Removed the unused variable.

pranith committed:
mem-cache: Add a tag mask to indexing policies

In some internal structures, a shorter tag is preferred to save storage.
This can be achieved by techniques like hashing the full tag to generate the
shortened tag or just ignoring the upper bits of the tag.

Adding a tag mask allows us to implement such techniques in gem5.

Change-Id: Ib339efd97bf33b706f93eb7f0eb5851ceb51afb6

pranith committed:
mem-cache: Use associative cache for VFT

The frequency table is better built using the generic cache library instead of the
AssociativeSet structure since we do not need the secure bit for this structure.

Change-Id: Ie3b6442235daec7b350c608ad1380bed58f5ccf4

pranith committed:
mem-cache: prefetch: Use associative cache for DCPT

The DCPT table is better built using the generic cache structure since we do not
need the secure field.

Change-Id: I8a4a8d3dab7fbc3bbc816107492978ac7f3f5934

pranith committed:
mem-cache: Derive tagged entry from cache entry

The tagged entry can be derived from the generic cache entry and add the secure
flag that it needs. This reduces code duplication.

Change-Id: I7ff0bddc40604a8a789036a6300eabda40339a0f

pranith commented:
> The code seems OK, but before we merge this I would like to understand the use case. When would you set the tag to _not_ be the entire address for an indexing policy?
> 
> If you could update the commit message with (1) the problem that you are solving and (2) a detailed explanation of why this is the solution to the problem, it will make it easier to review the code.

Thanks for the review. I updated the commit message and added a few other patches where I use this tag mask (the BTB patch). Thanks!

pranith commented:
I'll send a new PR.
